# E-Commerce Marketplace Analytics

## 1. Data Quality Check

Data quality is important because missing values, duplicate rows, and wrong data types can affect later analysis.

## 2. Importing the Dataset

The CSV files are loaded into Pandas from the raw data folder.

In [1]:
import pandas as pd

customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")
orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")
order_items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")
payments = pd.read_csv("../data/raw/olist_order_payments_dataset.csv")
reviews = pd.read_csv("../data/raw/olist_order_reviews_dataset.csv")
products = pd.read_csv("../data/raw/olist_products_dataset.csv")
sellers = pd.read_csv("../data/raw/olist_sellers_dataset.csv")
geolocation = pd.read_csv("../data/raw/olist_geolocation_dataset.csv")
category_translation = pd.read_csv("../data/raw/product_category_name_translation.csv")

## 3. Checking Missing Values

Missing values can affect analysis. Each table is checked before making any cleaning decision.

In [2]:
customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [3]:
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [4]:
order_items.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [5]:
payments.isnull().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [6]:
reviews.isnull().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [7]:
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [8]:
sellers.isnull().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

In [9]:
geolocation.isnull().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [10]:
category_translation.isnull().sum()

product_category_name            0
product_category_name_english    0
dtype: int64

### Missing Value Observation

The orders table has 160 missing approval dates, 1,783 missing carrier delivery dates, and 2,965 missing customer delivery dates. Reviews have many missing comment titles and messages because comments are optional. Products have missing values in some category, description, and dimension-related columns. These values are kept because they should not be removed or filled without a clear reason.

## 4. Checking Duplicates

Exact duplicate rows can affect results. Each table is checked below.

In [11]:
print("Customers:", customers.duplicated().sum())
print("Orders:", orders.duplicated().sum())
print("Order Items:", order_items.duplicated().sum())
print("Payments:", payments.duplicated().sum())
print("Reviews:", reviews.duplicated().sum())
print("Products:", products.duplicated().sum())
print("Sellers:", sellers.duplicated().sum())
print("Geolocation:", geolocation.duplicated().sum())
print("Category Translation:", category_translation.duplicated().sum())

Customers: 0
Orders: 0
Order Items: 0
Payments: 0
Reviews: 0
Products: 0
Sellers: 0
Geolocation: 261831
Category Translation: 0


### Duplicate Observation

No exact duplicate rows were found in eight tables. The geolocation table has 261,831 repeated rows. They are not deleted because many records can have the same ZIP-code prefix and location information.

## 5. Checking Data Types

The data types are checked to find columns that need conversion.

In [12]:
orders.dtypes

order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object

In [13]:
order_items.dtypes

order_id                   str
order_item_id            int64
product_id                 str
seller_id                  str
shipping_limit_date        str
price                  float64
freight_value          float64
dtype: object

In [14]:
payments.dtypes

order_id                    str
payment_sequential        int64
payment_type                str
payment_installments      int64
payment_value           float64
dtype: object

In [15]:
reviews.dtypes

review_id                    str
order_id                     str
review_score               int64
review_comment_title         str
review_comment_message       str
review_creation_date         str
review_answer_timestamp      str
dtype: object

In [16]:
products.dtypes

product_id                        str
product_category_name             str
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

In [17]:
customers.dtypes

customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str
dtype: object

### Date Conversion

The date columns were loaded as text. They are converted to datetime format. Missing dates will stay as `NaT`.

In [18]:
orders["order_purchase_timestamp"] = pd.to_datetime(
    orders["order_purchase_timestamp"]
)
orders["order_approved_at"] = pd.to_datetime(
    orders["order_approved_at"]
)
orders["order_delivered_carrier_date"] = pd.to_datetime(
    orders["order_delivered_carrier_date"]
)
orders["order_delivered_customer_date"] = pd.to_datetime(
    orders["order_delivered_customer_date"]
)
orders["order_estimated_delivery_date"] = pd.to_datetime(
    orders["order_estimated_delivery_date"]
)

In [19]:
order_items["shipping_limit_date"] = pd.to_datetime(
    order_items["shipping_limit_date"]
)

reviews["review_creation_date"] = pd.to_datetime(
    reviews["review_creation_date"]
)
reviews["review_answer_timestamp"] = pd.to_datetime(
    reviews["review_answer_timestamp"]
)

### Basic Numeric Type Check

Important price, freight, payment, and installment columns are checked to confirm that they are numeric.

In [20]:
print("Price:", order_items["price"].dtype)
print("Freight Value:", order_items["freight_value"].dtype)
print("Payment Value:", payments["payment_value"].dtype)
print("Payment Installments:", payments["payment_installments"].dtype)

Price: float64
Freight Value: float64
Payment Value: float64
Payment Installments: int64


### Verify Date Conversion

In [21]:
orders[
    [
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
].dtypes

order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

In [22]:
print("Shipping Limit Date:", order_items["shipping_limit_date"].dtype)
print("Review Creation Date:", reviews["review_creation_date"].dtype)
print("Review Answer Timestamp:", reviews["review_answer_timestamp"].dtype)

Shipping Limit Date: datetime64[us]
Review Creation Date: datetime64[us]
Review Answer Timestamp: datetime64[us]


### Saving Processed Data

Only the tables with converted date columns are saved in the processed data folder.

In [23]:
orders.to_csv("../data/processed/orders_clean.csv", index=False)
order_items.to_csv("../data/processed/order_items_clean.csv", index=False)
reviews.to_csv("../data/processed/reviews_clean.csv", index=False)

## 6. Data Quality Summary

- Eight tables have no exact duplicate rows.
- Geolocation has 261,831 repeated rows, but they were not deleted.
- Missing values were found in orders, reviews, and products.
- Order, shipping, and review date columns were converted to datetime.
- Price, freight, payment value, and installments are already numeric.
- No major issue was found that blocks the next project phase.

## 7. Conclusion

The data quality checks were completed before moving to business analysis. Missing values, duplicate records, and data types were reviewed. The needed date columns were converted to datetime format. The dataset is now ready for the next stage.